### ISSUE TITLE
Cheat Finder - Does it search all possible memory addresses?

### ISSUE THREAD

**Issue by Pr3ttyF1y** [byuu/bsnes#237]
This may or may not be an issue with BSNES, but I'm on a personal crusade to make good cheats for SNES Doom and while I've been successful in implementing infinite ammo and infinite health, I have hit a snag in granting access to all weapons.

In Cheat Engine x64, I can find the single byte memory address that controls your weapons inventory and can manipulate it on the fly in BSNES. However, I cannot find the corresponding memory address using the Cheat Finder. Cheat Engine finds 3 separate bytes that track with weapon inventory changes and while I can find 2 of those bytes in BSNES, the 2 found by both applications do not directly affect your weapons inventory if changed. However, the 3rd address does directly change your weapons inventory and that is the one I cannot find in BSNES.

So for this address, the byte starts out set to 03 when you start the game which is just your fist/pistol and then changes to 07 when you have fist/pistol/shotgun. A value of FF gives you all weapons and values in between give you various combinations of weapons.

I'm not an expert on pointers but I understand that could be a possibility. However, when I can make changes to the single byte address in Cheat Engine and see direct results in BSNES, I'm thinking that I should be able to find that corresponding address in BSNES, but I have not been able to. Therefore, I began to wonder if the Cheat Finder in BSNES is looking at every possible address (e.g., including expanded ram on the game PCB, etc.) or not. I'm sure Randy Linden did some insane things to get his 'Reality Engine' running on the SNES (I mean, could he stuff variables into the SPC700's ram?), so I was just wondering how extensive if BSNES's cheat search function.

Please close this issue if it's outside the realm of what should be posted in these issues. I just wasn't sure where else I could raise the question. Thanks - PF

**Comment by Kawa-oneechan** [comment]
For clarity's sake, you described a _bit field_ stating which weapons you have. What exactly is the _location_ of this value any way?

**Comment by Pr3ttyF1y** [comment]
[Here](https://i.ibb.co/KmF4PNz/doom.png) is an image of both the BSNES Cheat Finder and Cheat Engine windows with the values both found.

If you would be interested in confirming this search, I uploaded a [save file](https://www.sendspace.com/file/oaggnn) where you're standing right in front of the shotgun in E1M1 in the USA release of Doom.

**Comment by Kawa-oneechan** [comment]
Cheat Engine works with the _entire address space_ used by the target process (that is, BSNES). BSNES' Cheat Finder works with _only the emulated system's_. Thus, everything is offset and the search results are only valid for this particular version of BSNES.

**Comment by Pr3ttyF1y** [comment]
Absolutely understood. The addresses in the BSNES Cheat Finder stay static as that is the virtual SNES' memory while Cheat Engine is looking at the allocation of memory to the BSNES executable.

However, I ran replicate this search ad infinitum with any version of BSNES and I always get two addresses that correspond to your current weapons assortment in Doom. And in Cheat Engine I always get three addresses.

I understand the concept of offsets, but that I was noting that the two addresses in green (which I believe means that those addresses are static to that version of the BSNES executable) without a doubt correspond to the values that I find in the BSNES Cheat Editor. They're both 31 bytes part, just offset, like you said, because one is the SNES' memory map and the other is that of my PC running bsnes.exe.

My main hang up is that 3rd value that I get in Cheat Engine that directly controls which weapons you have. The other two sets of addresses that correspond with each other definitely change to correspond/match to which weapons you are currently holding, but they do not seem to actually control which weapons you are holding.

Thus, my question relates to that 3rd and final address in Cheat Engine. Is it part of the SNES' memory map that the Cheat Finder in BSNES is not able to find (i.e., it's not looking in all possible memory locations) or is that 3rd address some sort of abstraction internal to BSNES that is a product of emulation, but not part of the SNES memory map?

The only reason I'm raising the question is that value definitely will changes which weapons you have on the fly, I just can't find it in the Cheat Finder to actually add it as a PAR code.

**Comment by Kawa-oneechan** [comment]
"my guess: The mysterious third address refers to SRAM (or some chip-internal RAM, like SA-1 IRAM/BWRAM), but cheat finder only searches WRAM." — Alcaro on the discord.

Even though it's set to "all" in the screenshot, but who knows. SRAM *is* not just for battery-backed savegames, after all, the S is for Static.

**Comment by Pr3ttyF1y** [comment]
Thank you very much for reaching out. I appreciate it - and it has been [very productive](https://i.ibb.co/kcBd4rx/Doom-USA-001.png)

I got my code: 7001d9=ff

It ends up MAME's cheat search plugin is under the same constraints as BSNES. ZSNES' cheat search is even more limited.

The last time I used SNES9x, I don't think it had a cheat search, but it seems that 1.60 does and it seems like it searches outside of the SNES-side memory map. However, it doesn't seem quite accurate. it picked up on an address in SRAM listed as s001d9XX. Now, I'm familiar with the 1D9 as that's what Cheat Engine pointed me towards when I traced the 3rd value back to a pointer. However, SNES9x added the code as 8001d9=ff, which did not work. 

However, after looking at the manifest for DOOM in BSNES and making an educated guess, I ended up with 7001d9=ff that does work.

All in all, a bit convoluted, but I'm going to close out this issue. I have my answer regarding what BSNES provides (which I'm very thankful for) as well as alternative/combination mechanisms to achieve what I'm interested in achieving. Thanks again!
